service/frontend/handler.go

```diff
@@ -587,11 +583,25 @@ func (wh *WorkflowHandler) GetWorkflowExecutionHistory(
 			DomainUUID: common.StringPtr(info.ID),
 			Execution:  getRequest.GetExecution(),
 		})
-		if err != nil {
-			return nil, wrapError(err)
+		if err == nil {
+			token.nextEventID = response.GetEventId()
+			token.runID = response.GetRunId()
+		} else {
+			if _, ok := err.(*gen.EntityNotExistsError); !ok || !getRequest.GetExecution().IsSetRunId() {
+				return nil, wrapError(err)
+			}
+			// It is possible that we still have the events in the table even though the mutable state is gone
+			// Get the nextEventID from visibility store if we still have it.
+			visibilityResp, err := wh.visibitiltyMgr.GetClosedWorkflowExecution(&persistence.GetClosedWorkflowExecutionRequest{
```

Comment:
This path will never be hit by decision workers. If there is a decision, the workflow is not deleted. Also, they get a token from the PollForDecisionTask call anyway so we know what the nextEventId is from the token.